Update RELEASE.md (#7903)
paritytech/polkadot-sdk: docs/RELEASE.md

# Release

The outputs of a stable release are:

- The binaries:
   - `polkadot`
   - `polkadot-execute-worker`
   - `polkadot-prepare-worker`
   - `polkadot-parachain`
   - `polkadot-omni-node`
   - `chain-spec-builder`
   - `frame-omni-bencher`

   built for the `x86_64-unknown-linux-gnu` and `aarch64-apple-darwin` targets. The gpg signatures and sha256 checksums.

- The runtimes for Westend and its system parachains
- The new crate versions published to `crates.io`.

> 

Reviewer: And docker images?

# Timeline
`Stable` releases are scheduled on a quarterly basis, usually by the end of the last month of each quarter. The exact schedule can be found on the [Release Registry](https://github.com/paritytech/release-registry/). It is possible to subscribe to a [calendar link](https://raw.githubusercontent.com/paritytech/release-registry/main/releases-v1.ics) to have it in your personal calendar.

> Each stable release is supported for a period of one year from it's first release. For example, `Polkadot stable2412` was released on `2024-12-17` and its end of life is set to `2025-12-16`.

Reviewer 2: ```suggestion
Each stable release is supported for a period of one year from its first release. For example, `Polkadot stable2412` was released on `2024-12-17` and its end of life is set to `2025-12-16`.
```

During this period, each stable release is updated with patch releases, which are scheduled on a monthly basis and contain fixes for any bugs that may be found.

> ℹ️ Note: the binaries and runtimes (if needed) are only provided for the latest `stable' release, for the previous releases only the crates.io release takes place.

Reviewer: ```suggestion
ℹ️ Note: the binaries and runtimes (if needed) are only provided for the latest `stable` release, for the previous releases only the crates.io release takes place.
```

This three month period between `stable` releases includes a 1.5 month QA period. This means that for each upcoming `stable` release, the branch from which that release will be made is created 1.5 months before the release date. This time is used to test the upcoming release candidate and find any issues that may arise with publishing crates to crates.io, binary artifacts and template synchronisation before the final release. The findings should be fixed and backported to the release branch. During this time, multiple release candidates may be built and rolled out.

# Setup

> We have two branches: `master` and `stableYYMM`. `master` is the main development branch where normal Pull Requests are

Reviewer: Actually more than two now 😅

Author: Yep, this is more like a general concept, but I'll extend it😄

opened. Developers need to mostly only care about this branch.
The `stableYYMM` branch contains a version of the code that is ready to be released. Its contents should be always audited.

> Merging to it is restricted to [Backports](#backports). Each `stableYYMM` branch corresponds to the corresponding stable release, which is in the maintenance period.

Reviewer: Or in QA period, or?

## Node

The versioning of the Polkadot node is done most of the time by only incrementing the `minor` version. The `major`
version is only bumped for special releases and the `patch` is used for a patch release that happens every month and fixes found issues.
The node version is not following SemVer. This means that the version doesn't express if there are any breaking changes in the CLI interface or similar.
The node version is declared in the [`NODE_VERSION`](https://paritytech.github.io/polkadot-sdk/master/polkadot_node_primitives/constant.NODE_VERSION.html)
variable.

## Westend

> For these networks, in addition to incrementing the `Cargo.toml` version we also increment the `spec_version` and

Reviewer: ```suggestion
For the Westene testnet, in addition to incrementing the `Cargo.toml` version we also increment the `spec_version` and
```

sometimes the `transaction_version`. The spec version is also following the node version. Its schema is: `M_mmm_ppp` and
for example `1_002_000` is the node release `1.2.0`. This versioning has no further meaning, and is only done to map

> from an on chain `spec_version` easily to the release in this repository.

Reviewer: Reading this makes me think that it could be better to change our Westend versioning, but that is for another day.

**From `stable` to `master`**

> Backports to `master` only happen after a `stable` or `patch` release has been made for the current stable release, and include node version and spec version bumps, plus reorganizing the prdoc files (they should go into the appropriate release folder under the [prdoc](./proc) folder).

Reviewer: ```suggestion
Backports to `master` only happen after a `stable` or `patch` release has been made for the current stable release, and include node version and spec version bumps, plus reorganizing the prdoc files (they should go into the appropriate release folder under the [prdoc](./proc) folder).  
This is done by the release team to keep things organized. Developers need not to care about such backports.
```

### Steps

1. Developer opens a Pull Request with changed crates against `master`.
2. They mention the type of the bump of all changed crates according to SemVer in the prdoc file attached to the PR. Note that this includes any crates that expose the changed
   behaviour in their *public API* and also transitive dependencies for whom the same rule applies.
3. The bump itself happens during the release and is done by a release engineer using the [parity-publish](https://github.com/paritytech/parity-publish) tool.

## Stable Release

Cadence: every 3 months for new `stable` releases and monthly for existing `stables`. Responsible: Release Team.

### Steps to execute a new stable release

From the main polkadot-sdk repository in the paritytech org:

1. On the cut-off date, create a new branch with the name `satbleYYMM` using [Branch-off stable flow](/.github/workflows/release-10_branchoff-stable.yml)
2. Create a new rc tag from the stable branch using [RC Automation flow](/.github/workflows/release-11_rc-automation.yml)

> From the forked polkadot-sdk repository in the [paritytech-release org](https://github.com/paritytech-release/polkadot-sdk/actions):

Reviewer 3: I would add a 10th step regarding syncing templates, which are user facing. Similar for the latest stable release latest patch.

1. Sync the forks before continuing with the release using [Sync the forked repo with the upstream](https://github.com/paritytech-release/polkadot-sdk/actions/workflows/fork-sync-action.yml)
2. To build binaries trigger [Release - Build node release candidate](/.github/workflows/release-20_build-rc.yml)
3. When an rc build is ready to trigger [Release - Publish draft](/.github/workflows/release-30_publish_release_draft.yml) to create a new release draft for the upcoming rc
4. When the release is finalized and ready to go, publish crates using `parity-publish` tool and push changes to the release branch
5. Repeat steps 1-3 to prepare the rc
6. Trigger [Release - Promote RC to final candidate on S3](/.github/workflows/release-31_promote-rc-to-final.yml) to have it as a final rc on the S3
7. Publish deb package for the `polakdot` binary using [Release - Publish polakdot deb package](/.github/workflows/release-40_publish-deb-package.yml)
8. Adjust the release draft and publish release on the GitHub.
9. Publish docker images using [Release - Publish Docker Image](/.github/workflows/release-50_publish-docker.yml)

> 

Reviewer: Comms stuff is handled elsewhere?

Author: Sorry, did not get that🙈

From the main polkadot-sdk repository in the paritytech org:

1. Synchronize templates using [Synchronize templates](/.github/workflows/misc-sync-templates.yml)

> 

Reviewer: And also need to update the release registry

3. Audit happens with priority.
4. It is merged into `master`.
5. Dev adds the `A4-needs-backport` label.
6. It is automatically back-ported to `stable` and merged by a release engineer.
7. The fix will be released in the next *Stable patch* release. In urgent cases, a release can happen earlier.

> The release itself is similar to [the new stable release](#steps-to-execute-a-new-stable-release) process without the branching-off step, as the branch already exists and depending on the patch (whether it is for the current `stable` release or one of the previous ones) the binary build can be skipped and only crates and GitHub publishing is done.

Reviewer 2: dumb question: why do we skip binaries release for bug/security fixes?

Author: There was a discussion about it some time ago in the release coordination chat and on the forum: https://forum.parity.io/t/polkadot-binary-releases/2429/5
If there is something critical needed to be fixed, we will always build and attach those to the release.
